datafusion/functions/src/regex/regexpcount.rs

```diff
@@ -541,42 +541,45 @@ where
 
                     let pattern =
                         compile_and_cache_regex(regex, flags, &mut regex_cache)?;
-                    count_matches(value, &pattern, start)
+                    count_matches(value, pattern, start)
                 })
                 .collect::<Result<Vec<i64>, ArrowError>>()?,
             )))
         }
     }
 }
 
-fn compile_and_cache_regex(
-    regex: &str,
-    flags: Option<&str>,
-    regex_cache: &mut HashMap<String, Regex>,
-) -> Result<Regex, ArrowError> {
-    match regex_cache.entry(regex.to_string()) {
-        Entry::Vacant(entry) => {
-            let compiled = compile_regex(regex, flags)?;
-            entry.insert(compiled.clone());
-            Ok(compiled)
-        }
-        Entry::Occupied(entry) => Ok(entry.get().to_owned()),
+fn compile_and_cache_regex<'a>(
+    regex: &'a str,
+    flags: Option<&'a str>,
+    regex_cache: &'a mut HashMap<String, Regex>,
```

Comment:
Is `HashMap<&str, Regex>` much better for avoiding clone?

Reply:
Good point. I tried to adjust it but then the keys should live at least as long as HashMap, so I couldn't make it compile. regexp_replace also uses String as HashMap keys so I think it's fine. However, I removed one unnecessary clone by changing signature of compile_regex func.